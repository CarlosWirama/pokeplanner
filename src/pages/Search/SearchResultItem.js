import React from 'react';
import PropTypes from 'prop-types';
import {
  StyledListItem,
  PokemonImage,
  Texts,
  Name,
  Types,
  Type,
} from './SearchResultItem.styled.js';

export default function SearchResultItem({
  listItem: { name, types },
  onClick,
}) {
  return (
    <StyledListItem button onClick={onClick}>
      <PokemonImage>ssd</PokemonImage>
      <Texts>
        <Name>{name}</Name>
        <Types>
          {types.map((type, index) => (
            <Type key={index}>{type}</Type>
          ))}
        </Types>
      </Texts>
    </StyledListItem>
  );
}

SearchResultItem.propTypes = {
  listItem: PropTypes.shape({
    kantoDex: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    types: PropTypes.arrayOf(
      PropTypes.string.isRequired
      ).isRequired,
    }).isRequired,
  onClick: PropTypes.func,
};

SearchResultItem.defaultProps = {
  onClick: () => {},
};
