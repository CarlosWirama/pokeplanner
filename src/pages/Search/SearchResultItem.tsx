import React from 'react';
import PropTypes from 'prop-types';
import PokemonInfo from '../../common/components/PokemonInfo';
import { ResultItem } from './SearchResultItem.styled';

export default function SearchResultItem({
  listItem: { name, types, form },
  onClick,
}) {
  return (
    <ResultItem onClick={() => onClick(name, form)}>
      <PokemonInfo
        name={name}
        types={types}
        form={form}
        titleColor="white"
      />
    </ResultItem>
  );
}

SearchResultItem.propTypes = {
  listItem: PropTypes.shape({
    name: PropTypes.string.isRequired,
    types: PropTypes.arrayOf(
      PropTypes.string.isRequired
    ).isRequired,
    form: PropTypes.string,
  }).isRequired,
  onClick: PropTypes.func,
};

SearchResultItem.defaultProps = {
  onClick: () => {},
};
