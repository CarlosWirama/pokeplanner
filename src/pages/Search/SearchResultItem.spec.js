import React from 'react';
import renderer from 'react-test-renderer';
import SearchResultItem from './SearchResultItem';

const pokemonListItemMock = {
  kantoDex: '001',
  name: 'Bulbasaur',
  types: ['Grass', 'Poison'],
};

describe('SearchResultItem', () => {
  const onClickMock = jest.fn();

  // TODO dunno how to test this for now
  // it('should throw error when listItem prop is missing', () => {

  it('should render listItem props', () => {
    // TODO enzyme mount
    expect(0).toEqual(1);
  });

  // render the component here and reuse in below tests
  const component = renderer.create(
    <SearchResultItem
      listItem={pokemonListItemMock}
      onClick={onClickMock}
    />,
  );
  let tree = component.toJSON();

  it('should render without fail', () => {
    expect(tree).toMatchSnapshot();
  });


  it('should handle click correctly', () => {
    // manually trigger the callback
    tree.props.onClick();

    // check whether the function successfully called
    expect(onClickMock.mock.calls.length).toBe(1);

    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  const THIS_MANY_CLICKS = 20;
  it(`should handle ${THIS_MANY_CLICKS} clicks correctly`, () => {
    // clear previous history
    onClickMock.mockClear();

    // trigger THIS_MANY_CLICKS times
    for(let i = 0; i < THIS_MANY_CLICKS; i++) {
      tree.props.onClick();
    }

    // check whether the function successfully called
    expect(onClickMock.mock.calls.length).toBe(THIS_MANY_CLICKS);

    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
