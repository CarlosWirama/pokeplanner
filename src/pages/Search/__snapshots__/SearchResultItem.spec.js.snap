// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultItem should handle 20 clicks correctly 1`] = `
<div
  className="MuiButtonBase-root-13 MuiListItem-root-1 MuiListItem-default-4 MuiListItem-gutters-9 MuiListItem-button-10 sc-bdVaJa cPqfWC"
  onBlur={[Function]}
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
>
  <div
    className="sc-bwzfXH fVcXGd"
  >
    ssd
  </div>
  <div
    className="sc-htpNat hHtpcc"
  >
    <div
      className="sc-bxivhb jxkoxB"
    >
      Bulbasaur
    </div>
    <div
      className="sc-ifAKCX hrWLvt"
    >
      Grass
      Poison
    </div>
  </div>
  <span
    className="MuiTouchRipple-root-16"
  />
</div>
`;

exports[`SearchResultItem should handle click correctly 1`] = `
<div
  className="MuiButtonBase-root-13 MuiListItem-root-1 MuiListItem-default-4 MuiListItem-gutters-9 MuiListItem-button-10 sc-bdVaJa cPqfWC"
  onBlur={[Function]}
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
>
  <div
    className="sc-bwzfXH fVcXGd"
  >
    ssd
  </div>
  <div
    className="sc-htpNat hHtpcc"
  >
    <div
      className="sc-bxivhb jxkoxB"
    >
      Bulbasaur
    </div>
    <div
      className="sc-ifAKCX hrWLvt"
    >
      Grass
      Poison
    </div>
  </div>
  <span
    className="MuiTouchRipple-root-16"
  />
</div>
`;

exports[`SearchResultItem should render without fail 1`] = `
<div
  className="MuiButtonBase-root-13 MuiListItem-root-1 MuiListItem-default-4 MuiListItem-gutters-9 MuiListItem-button-10 sc-bdVaJa cPqfWC"
  onBlur={[Function]}
  onClick={[MockFunction]}
  onContextMenu={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
>
  <div
    className="sc-bwzfXH fVcXGd"
  >
    ssd
  </div>
  <div
    className="sc-htpNat hHtpcc"
  >
    <div
      className="sc-bxivhb jxkoxB"
    >
      Bulbasaur
    </div>
    <div
      className="sc-ifAKCX hrWLvt"
    >
      Grass
      Poison
    </div>
  </div>
  <span
    className="MuiTouchRipple-root-16"
  />
</div>
`;
